### ISSUE TITLE
crossfilter.remove with predicate is not always calling reduceRemove 

### ISSUE THREAD

**Issue by betalb** [crossfilter/crossfilter#89]
Somewhere in 1.4 developement phase, group functionallity was drastically changed and reduceAdd is called for every record, but remove may be skipped leading to failure of following test.

```
"remove with predicate and reduceRemove": {
  "removed records call reduceRemove on filtered records": function() {
    var data = crossfilter();
    var fooDimension = data.dimension(function(d) { return d.foo; });
    data.add([{foo: 1}, {foo: 2}, {foo: 3}]);

    var allFilteredVsTotal = data.groupAll().reduce(
      function(result, d, nf) {
        if (nf) {
          result.total += 1;
          result.filtered += 1;
        } else {
          result.filtered += 1;
        }

        return result;
      },
      function(result, d, nf) {
        if (nf) {
          result.total -= 1;
          result.filtered -= 1;
        } else {
          result.filtered -= 1;
        }

        return result;
      },
      function() {
        return {
          total: 0,
          filtered: 0
        };
      }
    );

    assert.deepEqual(allFilteredVsTotal.value(), {
      total: 3,
      filtered: 3
    });

    // Apply filter
    fooDimension.filter(2);
    assert.deepEqual(allFilteredVsTotal.value(), {
      total: 3,
      filtered: 1,
    });

    data.remove(function (d) {
      return d.foo === 3;
    });
    assert.deepEqual(allFilteredVsTotal.value(), {
      total: 2,
      filtered: 1,
    });
  }
}
```

Considering that, wanted to highlight few points:
* reading rationale for changing original behaviour, I can't understand why this can't be implemented as top level API, not group level. Special type of group that is not affected by dimension filters
* Performance will be affected if reducers are **heavy**. Previously there was possibility to apply filter that reduces cardinality and then create group. Now performance drop can be considerable
* If dataset contains values close to `Number.MAX_VALUE` or `Number.MIN_VALUE`, previously there was possibility to avoid overflow in cases when you slice your dataset using required filters which reduce cardinality, now there is no way to do this
* Floating point precision errors may become more evident. See example below, it illustrates datasets, consisting of 100,001 records [A, B × 100,000]. Previously there was possibility to avoid them by creating group after applying filter, now there is no way to do this

Old crossfilter
```
var A = 42.90;
var B = 0.001;
for (var i = 0; i < 20000; i += 1) {
  A += B;
}
console.log(A); // 62.89999999995339
```
New crossfilter
```
var A = 42.90;
var B = 0.001;
for (var i = 0; i < 100000; i += 1) {
  A += B;
}
for (var i = 0; i < 80000; i += 1) {
  A -= B;
}
console.log(A) // 62.89999999995337
```

This is not causing big issues, except when all records are filtered
```
var A = 42.90;
var B = 0.001;
for (var i = 0; i < 100000; i += 1) {
  A += B;
}
for (var i = 0; i < 100000; i += 1) {
  A -= B;
}
console.log(A) // 42.899999999999984
```

**Comment by esjewett** [comment]
The failure of that test function is a problem. Is the failure only when `remove` is called with a predicate function or is it more general?

Regarding rationale, the idea is to allow groups access to the full data lifecycle. Implementing a top level method would indeed address one use-case (seeing the unfiltered counts), but it wouldn't address related use-cases that we can't predict (the unfiltered counts for a specific subset of records).

I do see what you are saying about the performance impact in scenarios where you are looking at adding a lot of data. There was some extended discussion regarding the performance impact during the initial change. The thought, if I remember correctly, was that in the Crossfilter use-case, we expect that users will be able to change the current filter and that reducers should run fast enough to allow this to happen smoothly. As such, I didn't anticipate this being used in a way where we are adding data at a volume that would overwhelm the ability of reducers to keep up. I do think there is room for improvement here though. I would especially be interested in proposals for doing this by setting a flag on group creation and executing a different code path in these cases. Such a change would probably have to wait for a 2.0 release and I think it would probably be something we would want to do after a significant refactor, as the current organization of Crossfilter is not friendly to alternative code paths.

Regarding floating point error, this is a problem, but I generally recommend against doing FP operations in reducers. I'm sure it's sometimes unavoidable, but I actually haven't run into such a situation. Do you have an example?

**Comment by betalb** [comment]
The test fails only for `remove` with predicate, because only in this case filtered record can be removed.

In this case reduce remove was already called during data ingestion.

It looks like in this situation one of the options will be calling reduce add (with nf = false) and then reduce remove (with nf = true).

**Comment by esjewett** [comment]
Ok - sorry, I haven't been able to look at issues again since your last response. I'll try to take a shot at fixing this.

**Comment by thatcort** [comment]
I'm seeing erroneous behavior on adds too. I have a crossfilter.groupAll.reduce function. The add function is called for far fewer items if I incrementally change the applied filter range, than if I change it all at once. I could try to create a test case, but there doesn't seem to be much movement here.

**Comment by gordonwoodhull** [comment]
Hi @thatcort, this issue is referring to `crossfilter.remove`, which is unrelated to the ordinary `reduceRemove` called when a filter changes.

If you change a brush only slightly you should see much fewer `reduceRemove` events than if you make a big change, because the changes are made incrementally (crossfilter is designed to make that case really fast).

If you mean that fewer reduceRemove happen *total*, over the same total change to the brush, then that would be a bug, but I would be doubtful without seeing some evidence, because the filter and reduce functions have been the same for a long time and have gotten heavy testing. In contrast, the remove function here in this issue is relatively new.

**Comment by thatcort** [comment]
Yes, I'm seeing different results overall. I'll try to create a testcase and submit a new bug.

**Comment by betalb** [comment]
Just small update FYI

I've solved floating point issues by applying [Kahan summation algorithm](https://en.wikipedia.org/wiki/Kahan_summation_algorithm)
